servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpPayloadDiscardCleanerServiceFilter.java

```diff
@@ -0,0 +1,136 @@
+import io.servicetalk.http.api.HttpLifecycleObserver;
+import io.servicetalk.http.api.HttpRequestMetaData;
+import io.servicetalk.http.api.HttpResponseMetaData;
+import io.servicetalk.http.api.HttpServiceContext;
+import io.servicetalk.http.api.StreamingHttpRequest;
+import io.servicetalk.http.api.StreamingHttpResponse;
+import io.servicetalk.http.api.StreamingHttpResponseFactory;
+import io.servicetalk.http.api.StreamingHttpService;
+import io.servicetalk.http.api.StreamingHttpServiceFilter;
+import io.servicetalk.http.api.StreamingHttpServiceFilterFactory;
+import io.servicetalk.transport.api.ConnectionInfo;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import static io.servicetalk.http.netty.HttpPayloadDiscardWatchdogServiceFilter.payloadPublisherKey;
+import static io.servicetalk.http.netty.HttpPayloadDiscardWatchdogServiceFilter.payloadSubscribedKey;
+
+final class HttpPayloadDiscardCleanerServiceFilter implements StreamingHttpServiceFilterFactory, HttpLifecycleObserver {
+
+    /**
+     * Instance of {@link HttpPayloadDiscardCleanerServiceFilter}.
+     */
+    static final StreamingHttpServiceFilterFactory INSTANCE = new HttpPayloadDiscardCleanerServiceFilter();
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(HttpPayloadDiscardCleanerServiceFilter.class);
+
+    private HttpPayloadDiscardCleanerServiceFilter() {
+        // Singleton
+    }
+
+    @Override
+    public StreamingHttpServiceFilter create(final StreamingHttpService service) {
+
+        final WatchdogHttpLifecycleObserver observer = new WatchdogHttpLifecycleObserver(this, false);
+
+        return new StreamingHttpServiceFilter(service) {
+            @Override
+            public Single<StreamingHttpResponse> handle(final HttpServiceContext ctx,
+                                                        final StreamingHttpRequest request,
+                                                        final StreamingHttpResponseFactory responseFactory) {
+                return observer.trackLifecycle(ctx, request, r -> delegate().handle(ctx, r, responseFactory));
+            }
+
+            @Override
+            public HttpExecutionStrategy requiredOffloads() {
+                return HttpExecutionStrategies.offloadNone();
+            }
+        };
+    }
+
+    private static class WatchdogHttpLifecycleObserver extends AbstractLifecycleObserverHttpFilter {
+        WatchdogHttpLifecycleObserver(final HttpLifecycleObserver observer, final boolean client) {
+            super(observer, client);
+        }
+    }
+
+    @Override
+    public HttpExchangeObserver onNewExchange() {
+
+        return new HttpExchangeObserver() {
+
+            @Nullable
+            private HttpRequestMetaData requestMetaData;
+
+            @Override
+            public HttpRequestObserver onRequest(final HttpRequestMetaData requestMetaData) {
+                this.requestMetaData = requestMetaData;
```

Comment:
Consider saving only `request.context()`

Reply:
Caching the whole metadata helps with debug logging the surrounding context - do you think it's worth the optimization still?

Reply:
I saw that you used it for logging and forgot to update my comment. Even though we filter headers by default,  potential risk there is that you can log sensitive request-target if path contains auth token.
It shouldn't matter what request caused it bcz if users have a problem, it's likely applicable for every request. And if they want more details, they can add correlation key in MDC.

Reply:
Ok thanks for the clarification - updated